modyn/trainer_server/internal/trainer/remote_downsamplers/remote_craig_downsampling.py

```diff
@@ -0,0 +1,196 @@
+from argparse import Namespace
+from typing import Any, Optional
+
+import numpy as np
+import torch
+from modyn.trainer_server.internal.trainer.remote_downsamplers.abstract_per_label_remote_downsample_strategy import (
+    AbstractPerLabelRemoteDownsamplingStrategy,
+)
+from modyn.trainer_server.internal.trainer.remote_downsamplers.deepcore_utils import submodular_optimizer
+from modyn.trainer_server.internal.trainer.remote_downsamplers.deepcore_utils.euclidean import euclidean_dist_pair_np
+from modyn.trainer_server.internal.trainer.remote_downsamplers.deepcore_utils.shuffling import _shuffle_list_and_tensor
+from modyn.trainer_server.internal.trainer.remote_downsamplers.deepcore_utils.submodular_function import (
+    FacilityLocation,
+)
+from modyn.trainer_server.internal.trainer.remote_downsamplers.deepcore_utils.submodular_optimizer import (
+    OPTIMIZER_CHOICES,
+)
+
+
+class RemoteCraigDownsamplingStrategy(AbstractPerLabelRemoteDownsamplingStrategy):
```

Comment:
Just for my understanding: Why do some downsamplers have their algorithm living in deepcore_utils (like kcenter_greedy), and others (like craig) have their algorithm here?

Reply:
CRAIG's selection algorithm is in deepcore utils ( ``` from modyn.trainer_server.internal.trainer.remote_downsamplers.deepcore_utils.submodular_function import FacilityLocation ``` ). All the selection algorithms are in a separate file, except for the trivial ones (like top-k or weighted sampling). This follows the design of Deepcore and makes sense since the same algorithm can be used for different strategies (ex, submodularFL is used by Craig and Submodular)